Compress draco 2

jcelerier opened the pull request:
This pull request starts to have something that looks like a working implementation (see https://github.com/KDAB/kuesa/pull/16/files?w=1 to get a diff without the whitespace changes - github really ought to have a nice button for this).

### What works
* Compressing and reloading these in Kuesa: 
  - https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0/Box/glTF
  - https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0/Cube/glTF
  - https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0/AnimatedCube/glTF
  - https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0/BoxAnimated/glTF
  - https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0/Duck/glTF
  - https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0/Buggy/glTF
  - The DodgeViper example

### What doesn't
* Original buffers aren't stripped of uncompressed data properly yet (it works if there is no interleaving of multiple bufferViews) - edit : actually it looks like this works (at least the interleaved glTF example works : https://github.com/KhronosGroup/glTF-Sample-Models/tree/master/2.0/BoxInterleaved ) ; I suppose that this is because once we're in the Qt3D world (as the mesh data are), there is no interleaving anymore ? 

* Currently, all the indices relative to the following bufferViews in the whole GLTF file (in known extensions) are readjusted. But if someone has a custom extension that also uses bufferViews to access data, we won't have any way to know that we have to adjust them and this will break the custom extension. (and we can't assume that it will be exposed as `"bufferView": 123` to try and replace them, it may be a different key like the ones used in `"primitives"`).

### What's missing
* tests, prettyfication, proper comments and documentation
* support for embedded data (depends on https://github.com/KDAB/kuesa/pull/25)

jcelerier commented:
@mkrus it's on top of current master now

mkrus commented:
guess this needs rebased on master once the parser context changes are merged

mkrus commented on tools/assetpipelineeditor/mainwindow.cpp:
```diff
@@ -337,6 +345,42 @@ void MainWindow::openFile()
     setFilePath(filePath);
 }
 
+void MainWindow::compressFile()
```
need error reporting if compressing fails, write files fails, etc

jcelerier replied:
I've added error reporting in the export gui, is it enough ?

er-vin replied:
I guess it'll do for now

mkrus commented on tests/auto/gltfexporter/tst_gltfexporter.cpp:
```diff
@@ -0,0 +1,80 @@
+    This file is part of Kuesa.
+
+    Copyright (C) 2018 Klarälvdalens Datakonsult AB, a KDAB Group company, info@kdab.com
+    Author: Mike Krus <mike.krus@kdab.com>
```
not me :)

jcelerier replied:
done

mkrus commented on tests/auto/gltfexporter/gltfexporter.pro:
```diff
@@ -0,0 +1,39 @@
+# This file is part of Kuesa.
+#
+# Copyright (C) 2018 Klarälvdalens Datakonsult AB, a KDAB Group company, info@kdab.com
+# Author: Mike Krus <mike.krus@kdab.com>
```
not me :)

jcelerier replied:
done

jcelerier commented on src/core/gltf2importer/dracocompressor_p.h:
```diff
@@ -0,0 +1,28 @@
+}
+namespace Kuesa {
+namespace Compression {
+struct DracoError : std::runtime_error {
```
done

jcelerier commented on src/core/gltf2importer/gltf2importer.h:
```diff
@@ -45,6 +45,7 @@ class KUESASHARED_EXPORT GLTF2Importer : public Qt3DCore::QNode
     Q_PROPERTY(QUrl source READ source WRITE setSource NOTIFY sourceChanged)
     Q_PROPERTY(Kuesa::GLTF2Importer::Status status READ status NOTIFY statusChanged)
     Q_PROPERTY(Kuesa::SceneEntity *sceneEntity READ sceneEntity WRITE setSceneEntity NOTIFY sceneEntityChanged)
+    Q_PROPERTY(Kuesa::GLTF2Context *context READ context)
```
done

jcelerier commented on src/quick/imports/kuesaplugin.cpp:
```diff
@@ -73,6 +74,7 @@ void KuesaPlugin::registerTypes(const char *uri)
 
     // Misc
     qmlRegisterType<Kuesa::GLTF2Importer>(uri, 1, 0, "GLTF2Importer");
+    qmlRegisterType<Kuesa::GLTF2Exporter>(uri, 1, 0, "GLTF2Exporter");
```
done (we discussed this a bit with @lemirep  IIRC)

jcelerier commented on tools/assetpipelineeditor/mainwindow.ui:
```diff
@@ -183,20 +186,25 @@
     <string>Ctrl+,</string>
    </property>
   </action>
+  <action name="actionSave">
```
ok, indeed save is not ideal especially considering that this is a fairly destructive operation - currently I save in a subfolder named "compressed". there could also be the possibility to set the compression ratio - draco seems to be tweakable per-attribute.

jcelerier replied:
done (changed to Export)

mkrus commented:
it's fine to the tool to access the private classes, they will all be release together

mkrus commented on tests/auto/gltfexporter/tst_gltfexporter.cpp:
```diff
@@ -4,7 +4,7 @@
     This file is part of Kuesa.
 
     Copyright (C) 2018 Klarälvdalens Datakonsult AB, a KDAB Group company, info@kdab.com
```
2019

mkrus commented on tests/auto/gltfexporter/gltfexporter.pro:
```diff
@@ -3,7 +3,7 @@
 # This file is part of Kuesa.
 #
 # Copyright (C) 2018 Klarälvdalens Datakonsult AB, a KDAB Group company, info@kdab.com
```
2019

mkrus commented on src/core/gltf2exporter/dracocompressor_p.cpp:
```diff
@@ -0,0 +1,275 @@
+#include "dracocompressor_p.h"
```
missing copyright header

jcelerier commented:
I rebased on master and added a dialog that will allow to set-up the export options, hope the design is correct (the bottom part reports the errors that may happen during compression) : 
![dialog](https://user-images.githubusercontent.com/2772730/51386615-57104480-1b23-11e9-8095-e198d147c6b7.png)

mkrus commented:
> I rebased on master and added a dialog that will allow to set-up the export options, hope the design is correct (the bottom part reports the errors that may happen during compression) :
> ![dialog](https://user-images.githubusercontent.com/2772730/51386615-57104480-1b23-11e9-8095-e198d147c6b7.png)

missing changes to `assetpipelineeditor.pro` to include new files

jcelerier commented:
Added the changes, sorry !

jcelerier commented:
> the OS will prompt you.

So, the only thing that may be dangerous is that you prompt the file chooser for a .gltf file, but it's not only the .gltf which will change but also the assets which will be stripped of data (and it cannot even be recovered since draco is lossy).
e.g. if for instance you save in the same folder but with a different name, you won't get prompted by the OS but the assets will silently change.

jcelerier commented:
I rebased on master, since GLTFContext is back to private and is needed for the export I added the exporter as a friend of the importer so that it can go and fetch it without having GLTFContext leaking in the public api. (if there are any better ways to do this please tell me !)

jcelerier commented:
Managed to fix the compression crash ! e.g. the dodgeViper does not crash anymore while compressing (it doesn't reload though, investigating.). Much more examples of the GLTF2 tests are able to compress and reload, e.g. here's a fairly quantized duck : 
![ducky](https://user-images.githubusercontent.com/2772730/51832791-f092df80-22f5-11e9-9c31-5227bb76a4a0.png)

lemirep commented on src/core/gltf2importer/gltf2exporter.cpp:
```diff
@@ -0,0 +1,372 @@
+
+                        // TODO the names aren't the correct ones, how to get them ?
+                        // e.g. vertexNormal instead of NORMAL
+                        // for the sake of testing, this dirty hack :
```
QAttribute::defaultPositionAttributeName, QAttribute::defaultNormalAttributeName, QAttribute::defaultColorAttributeName, QAttribute::defaultTextureCoordinateAttributeName, QAttribute::defaultTextureCoordinate1AttributeName, QAttribute::defaultTextureCoordinate2AttributeName,
QAttribute::defaultTangentAttributeName, QAttribute::defaultJointIndicesAttributeName, QAttribute::defaultJointWeightsAttributeName

jcelerier replied:
I should maybe squash my commits so that there is only the last version since there has been a bunch of back-and-forth,  what do you think ?

lemirep replied:
Actually that's fine. I found I could use the Files Changed tab to do the review which gives me the same as if all commits had been squashed. Please ignore comments which aren't relevant any more.

lemirep commented on src/core/gltf2importer/gltf2exporter.cpp:
```diff
@@ -235,23 +244,238 @@ class GLTF2DracoCompressor
+        for (int i = 0, n = m_bufferViews.size(); i < n; i++) {
+            if (indices.count(i) != 0) {
+                currentOffset++;
+            }
```
++currentOffset; {} not needed

lemirep commented on src/core/gltf2importer/gltf2exporter.cpp:
```diff
@@ -235,23 +244,238 @@ class GLTF2DracoCompressor
+        for (int i = 0; i < m_accessors.size(); i++) {
+            auto acc = m_accessors[i].toObject();
+
+            auto it = acc.find(GLTF2Import::KEY_BUFFERVIEW);
```
Probably nicer to use const QJsonValue v = acc.value(KEY_BUFFERVIEW); if (!v.isUndefined()) ...

lemirep commented on src/core/gltf2importer/gltf2exporter.cpp:
```diff
@@ -235,23 +244,238 @@ class GLTF2DracoCompressor
+            for (int p = 0; p < primitives.size(); p++) {
+                auto primitive = primitives[p].toObject();
+
+                auto ext_it = primitive.find(GLTF2Import::KEY_EXTENSIONS);
```
I'd go for the QJsonValue here as well as it improves readability, the iterator only makes sense to avoid multiple lookups or to delete the entry.

lemirep commented on src/core/gltf2importer/dracocompressor_p.cpp:
```diff
@@ -149,11 +157,15 @@ QVector<T> createIndicesVectorFromAttribute(const Qt3DRender::QAttribute &attrib
     indicesVector.resize(attribute.count());
 
     if (byteStride == 0)
+    {
```
We usually keep the opening { on the same line as the if statement

lemirep commented on src/core/gltf2importer/gltf2exporter.cpp:
```diff
@@ -403,16 +392,19 @@ class GLTF2DracoCompressor
-            if (!f.open(QIODevice::ReadWrite))
+            QFile source_f(m_basePath.absoluteFilePath(uri));
+            if (!source_f.open(QIODevice::ReadOnly))
+            {
```
{ should be on same line as if

lemirep commented on src/core/gltf2importer/gltf2exporter.cpp:
```diff
@@ -546,16 +543,37 @@ void GLTF2Exporter::setContext(GLTF2Context *context)
+        return {};
 #endif
 
+    if(source != target)
```
missing space after if, { should be on same line

lemirep commented on src/core/gltf2importer/gltf2exporter.cpp:
```diff
@@ -546,16 +543,37 @@ void GLTF2Exporter::setContext(GLTF2Context *context)
+        // Copy images
+        const auto images = rootObject[GLTF2Import::KEY_IMAGES].toArray();
+        for (const auto& val : images)
+        {
```
here too

lemirep commented on src/core/gltf2importer/gltf2exporter.cpp:
```diff
@@ -546,16 +543,37 @@ void GLTF2Exporter::setContext(GLTF2Context *context)
+            const auto& image = val.toObject();
+            auto uri_it = image.find(GLTF2Import::KEY_URI);
+            if(uri_it != image.end())
+            {
```
and here

lemirep commented on src/core/gltf2importer/gltf2exporter.cpp:
```diff
@@ -546,16 +543,37 @@ void GLTF2Exporter::setContext(GLTF2Context *context)
+            auto uri_it = image.find(GLTF2Import::KEY_URI);
+            if(uri_it != image.end())
+            {
+                QFileInfo destFile(target, uri_it->toString());
```
const QFileInfo

lemirep commented on src/core/gltf2importer/gltf2exporter.h:
```diff
@@ -61,7 +61,10 @@ public Q_SLOTS:
+    QJsonObject compress(
+            QDir sourceFolder,
+            QDir targetFolder,
+            QJsonObject rootObject);
```
const QJsonObject &

jcelerier replied:
I also pass the QDir as const& now, is it ok ?

lemirep commented on tools/assetpipelineeditor/mainwindow.cpp:
```diff
@@ -362,12 +362,18 @@ void MainWindow::compressFile()
+        auto srcdir = QFileInfo(gltfFile).dir();
+        srcdir.mkdir("compressed");
+
+        auto targetdir = srcdir;
```
const auto targetdir = srcdir.cd("compressed");

jcelerier replied:
does not exist anymore

lemirep commented on tools/assetpipelineeditor/mainwindow.cpp:
```diff
@@ -362,12 +362,18 @@ void MainWindow::compressFile()
+        targetdir.cd("compressed");
+
+        rootObject = exporter.compress(srcdir, targetdir, rootObject);
         if(rootObject.empty())
```
space after if

lemirep commented on tests/auto/gltfexporter/gltfexporter.pro:
```diff
@@ -0,0 +1,39 @@
+# This file is part of Kuesa.
+#
+# Copyright (C) 2018 Klarälvdalens Datakonsult AB, a KDAB Group company, info@kdab.com
+# Author: Mike Krus <mike.krus@kdab.com>
```
I don't think Mike actually wrote this :)

jcelerier replied:
likewise, in the latest commit it has already been fixed (https://github.com/KDAB/kuesa/pull/16/files#diff-1f8bf50859e4b422c66a957bfb568cafR6)

lemirep commented on tests/auto/gltfexporter/tst_gltfexporter.cpp:
```diff
@@ -0,0 +1,80 @@
+    This file is part of Kuesa.
+
+    Copyright (C) 2018 Klarälvdalens Datakonsult AB, a KDAB Group company, info@kdab.com
+    Author: Mike Krus <mike.krus@kdab.com>
```
Really nice of you :)

lemirep commented on src/core/gltf2importer/gltf2exporter.cpp:
```diff
@@ -99,18 +99,25 @@ class GLTF2DracoCompressor
-        m_root[GLTF2Import::KEY_IMAGES] = std::move(m_images);
+        auto replace_json = [&] (QJsonObject& m_root, const QLatin1String& k, QJsonArray& arr) {
+            auto it = m_root.find(k);
+            if(it != m_root.end() && arr.empty())
```
spaces between if and (

lemirep commented on src/core/gltf2importer/gltf2exporter.cpp:
```diff
@@ -559,18 +568,22 @@ QJsonObject GLTF2Exporter::compress(
-        for (const auto& val : images)
+        auto images_it = rootObject.find(GLTF2Import::KEY_IMAGES);
+        if(images_it != rootObject.end())
         {
```
{ on if line

lemirep commented on tests/auto/gltfexporter/tst_gltfexporter.cpp:
```diff
@@ -33,43 +33,112 @@
 {
     Q_OBJECT
 
+    QDir setupTestFolder()
```
If would even go for a free standing function in an anonymous namespace

jcelerier replied:
done

lemirep commented on tools/assetpipelineeditor/exportdialog.cpp:
```diff
@@ -45,9 +38,24 @@ void ExportDialog::on_save()
-            QFileInfo(m_targetFile).dir());
+    m_targetFile = QFileDialog::getSaveFileName(this, tr("Select a folder to save to"), m_targetFile, tr("glTF 2.0 file (*.gltf)"));
+
+    QDir orig_dir = QFileInfo(m_originalFile).dir();
```
const ?

jcelerier replied:
done

lemirep commented on tools/assetpipelineeditor/exportdialog.cpp:
```diff
@@ -45,9 +38,24 @@ void ExportDialog::on_save()
+    m_targetFile = QFileDialog::getSaveFileName(this, tr("Select a folder to save to"), m_targetFile, tr("glTF 2.0 file (*.gltf)"));
+
+    QDir orig_dir = QFileInfo(m_originalFile).dir();
+    QDir target_dir = QFileInfo(m_targetFile).dir();
```
const ?

jcelerier replied:
done

lemirep commented on tools/assetpipelineeditor/exportdialog.cpp:
```diff
@@ -45,9 +38,24 @@ void ExportDialog::on_save()
+    QDir orig_dir = QFileInfo(m_originalFile).dir();
+    QDir target_dir = QFileInfo(m_targetFile).dir();
+    if(orig_dir.absolutePath() == target_dir.absolutePath())
+    {
```
if style

jcelerier replied:
done

lemirep commented on tools/assetpipelineeditor/exportdialog.cpp:
```diff
@@ -45,9 +38,24 @@ void ExportDialog::on_save()
+                    tr("Saving in the original folder will overwrite assets."),
+                    QMessageBox::Ok | QMessageBox::Cancel,
+                    QMessageBox::Cancel);
+        if(res != QMessageBox::Ok) {
```
space after if

jcelerier replied:
done

lemirep commented on src/core/gltf2exporter/gltf2exporter.h:
```diff
@@ -33,13 +33,50 @@
+class KUESASHARED_EXPORT GLTF2ExportConfiguration
+{
+public:
+    GLTF2ExportConfiguration();
```
why not have = default here directly?

lemirep commented on src/core/gltf2exporter/gltf2exporter.h:
```diff
@@ -33,13 +33,50 @@
+    GLTF2ExportConfiguration(GLTF2ExportConfiguration &&other);
+    GLTF2ExportConfiguration &operator=(const GLTF2ExportConfiguration &other);
+    GLTF2ExportConfiguration &operator=(GLTF2ExportConfiguration &&other);
+
```
Do you even need to define all of these if all are the default?

jcelerier replied:
actually no, I originally took inspiration from other Qt classes but they are not needed at all - removed.

lemirep commented on src/core/gltf2exporter/dracocompressor_p.cpp:
```diff
@@ -187,12 +187,12 @@ QVector<T> createIndicesVectorFromAttribute(const Qt3DRender::QAttribute &attrib
     if (byteStride == 0)
     {
         for (int i = 0, nbIndices = indicesVector.size(); i < nbIndices; ++i)
-            indicesVector[i] = static_cast<T>(data.data()[i * sizeof(T)]);
```
why not do const T *attributeData = static_cast<T *>(data.constData()); and in the loop indicesVector[i] = attributeData[i]; ?

lemirep commented on src/core/gltf2exporter/dracocompressor_p.cpp:
```diff
@@ -0,0 +1,312 @@
+                              int count)
+{
+    QByteArray outputBuffer;
+    outputBuffer.resize(count * vertexSize * sizeof(T));
```
vertexSize * sizeof(T) could probably cached into a variable since it's reused in several places

jcelerier replied:
done

lemirep commented on src/core/gltf2exporter/dracocompressor_p.cpp:
```diff
@@ -0,0 +1,312 @@
+    QVector<T> indicesVector;
+    indicesVector.resize(attribute.count());
+
+    if (byteStride == 0)
```
if / else style

jcelerier replied:
done (I changed this part so that there isn't any if / else since we can precompute what the actual stride will be)

lemirep commented on src/core/gltf2exporter/dracocompressor_p.cpp:
```diff
@@ -0,0 +1,312 @@
+    const auto data = attribute.buffer()->data().mid(byteOffset);
+
+    QVector<T> indicesVector;
+    indicesVector.resize(attribute.count());
```
could be written as QVector<T> indicesVector(attribute.count()); directly

jcelerier replied:
done (also is there a way to do it without initializing values ? considering that they will be set just afterwards)

lemirep replied:
I guess you could do QVector<T> indicesVector; indicesVector.reserve(attribute.count()); which should preallocate without initializing value. Then you need to use push_back to insert values though.

jcelerier replied:
hmm I think that it would be a bit less performant - let's keep it as ` QVector<T> indicesVector(static_cast<int>(attrCount))` :)

lemirep commented on src/core/gltf2exporter/dracocompressor_p.cpp:
```diff
@@ -0,0 +1,312 @@
+            continue;
+        }
+
+        qDebug() << "compressing: " << attribute->name() << attribute->vertexBaseType() << attribute->attributeType();
```
maybe use categorized logging instead?

jcelerier replied:
I've removed this qdebug, it was only to help me fix a bug - what kind of useful logging could be added instead ?

lemirep replied:
You could keep it but maybe log it into a dedicated category Kuesa::DracoCompressor or something?

jcelerier replied:
done

lemirep commented on src/core/gltf2exporter/dracocompressor_p.h:
```diff
@@ -0,0 +1,53 @@
+class GLTF2ExportConfiguration;
+namespace Compression {
+
+struct CompressedMesh {
```
{ should be on new line for structs and classes I believe

jcelerier replied:
hmm, when I run clang-format on that file (with `clang-format -i dracocompressor_p.h`)  it puts the brace here

er-vin replied:
yes, structs and classes have different rules. structs are treated like enums regarding that point.

jcelerier replied:
I added a destructor to that struct (in order to not require a draco header in that file) - should I keep it as a struct or change it to a class ?

jcelerier replied:
done

er-vin replied:
I'd say class in such a case. But I don't see the aforementioned destructor anyway.

lemirep commented on tools/assetpipelineeditor/exportdialog.cpp:
```diff
@@ -0,0 +1,88 @@
+    QDir orig_dir = QFileInfo(m_originalFile).dir();
+    QDir target_dir = QFileInfo(m_targetFile).dir();
+    if (orig_dir.absolutePath() == target_dir.absolutePath()) {
+        int res = QMessageBox::warning(
```
const int res

jcelerier replied:
done

jcelerier commented on src/core/gltf2exporter/dracocompressor_p.cpp:
```diff
@@ -0,0 +1,312 @@
+    }
+
+    auto bb = dracoMesh.ComputeBoundingBox();
+    qDebug() << bb.max_point()[0]
```
no, removed

jcelerier commented on tools/assetpipelineeditor/mainwindow.ui:
```diff
@@ -24,14 +24,15 @@
    <widget class="QMenu" name="menuFile">
     <property name="title">
-     <string>File</string>
+     <string>Fi&amp;le</string>
```
done

jcelerier commented on tools/assetpipelineeditor/mainwindow.ui:
```diff
@@ -61,7 +62,7 @@
   <widget class="QDockWidget" name="collectionBrowserDockWidget">
    <property name="windowTitle">
-    <string>Collection Browser</string>
+    <string>&amp;Collection Browser</string>
```
done for this, but where is the Camera ? I don't find it in the mainwindow form

jcelerier replied:
found it

jcelerier commented on tools/assetpipelineeditor/dracocompressor.cpp:
```diff
@@ -0,0 +1,166 @@
+
+        const auto attributeBuffer = attribute->buffer();
+        if (attributeBufferToDataBuffer.find(attributeBuffer) == std::end(attributeBufferToDataBuffer))
+            attributeBufferToDataBuffer[attributeBuffer] = createDracoBufferFromQBuffer(*attributeBuffer);
```
this code has since changed

jcelerier commented on tools/assetpipelineeditor/dracocompressor.cpp:
```diff
@@ -0,0 +1,166 @@
+void addFacesToMesh(const QVector<T> &indices, draco::Mesh &dracoMesh)
+{
+    for (int i, nbIndices = indices.size()/3; i < nbIndices; ++i)
+        dracoMesh.AddFace({static_cast<draco::IndexType<unsigned int, draco::PointIndex_tag_type_>>(indices[3*i+0]),
```
it seems that it does, this has been fixed

jcelerier commented on tools/assetpipelineeditor/dracocompressor.h:
```diff
@@ -0,0 +1,14 @@
+class QGeometry;
+};
+
+std::unique_ptr<draco::EncoderBuffer> compressMesh(const Qt3DRender::QGeometry &geometry);
```
I had named it "Compression" in a further commit, do you want me to change it to DracoCompressor ?

lemirep replied:
That could be a good idea, I suspect we might have other Compression classes that deal with Textures or other unrelated assets later on.

jcelerier replied:
done

lemirep commented:
Kuesa: WARNING: kuesa/src/core/gltf2exporter/dracocompressor_p.h does not have the "We mean it." warning
Kuesa: WARNING: kuesa/src/core/gltf2exporter/gltf2exporter_p.h does not have the "We mean it." warning

er-vin commented on src/core/gltf2exporter/gltf2exporter_p.h:
```diff
@@ -0,0 +1,115 @@
+class KUESASHARED_EXPORT GLTF2ExportConfiguration
+{
+public:
+    enum MeshAttribute { Position,
```
Please put Position on its own line

jcelerier replied:
done

er-vin commented on src/core/gltf2exporter/gltf2exporter_p.h:
```diff
@@ -0,0 +1,115 @@
+    GLTF2ExportConfiguration configuration() const;
+    void setConfiguration(const GLTF2ExportConfiguration &conf);
+
+    void setContext(GLTF2Import::GLTF2Context *context);
```
Context is private API so it's unlikely user code will be able to put one here (or we force them to do use kuesa-private). Are we sure this is required?
Ditto for the following method, this context class really feels like an implementation detail leaking through API.

jcelerier replied:
hmm, it is "only" needed to access the structures defined in meshparser_p.h - I don't see other easy ways right now (short of re-parsing the original gltf inside the gltfexporter, recreating the qt3d nodes, and operating on those instead).

jcelerier replied:
another solution would be to rewrite the dracocompressor file so that it isn't based on qt3d entities / attributes / etc. However this will break the day where the scene can be modified / combined / whatever, and since it is on the roadmap...

mkrus replied:
well GLTF2Exporter is also private so what's the issue?

er-vin replied:
Oh right, I was confused by the export macro... But indeed for some reason we don't have a KUESA_PRIVATE_EXPORT macro. Maybe that should be changed? (in a separate commit)

er-vin replied:
Now that it's introduced (thanks btw!), maybe it's time to start using it here?

er-vin commented on tests/auto/gltfexporter/tst_gltfexporter.cpp:
```diff
@@ -0,0 +1,198 @@
+/*
+    tst_gltfparser.cpp
```
Wrong filename?

jcelerier replied:
done

er-vin commented on tools/assetpipelineeditor/exportdialog.cpp:
```diff
@@ -0,0 +1,88 @@
+#include <QMessageBox>
+
+ExportDialog::ExportDialog(
+        Kuesa::GLTF2Exporter &exporter,
```
Parameters would be better on a single line.

jcelerier replied:
done

er-vin commented on tools/assetpipelineeditor/exportdialog.cpp:
```diff
@@ -0,0 +1,88 @@
+        Kuesa::GLTF2Exporter &exporter,
+        const QString &path,
+        QWidget *parent)
+    : QDialog(parent), ui(new Ui::ExportDialog), m_exporter(exporter), m_originalFile(path), m_targetFile(m_originalFile)
```
OTOH it'd be best to have one per line here.

jcelerier replied:
done

er-vin commented on tools/assetpipelineeditor/exportdialog.h:
```diff
@@ -0,0 +1,33 @@
+    ~ExportDialog();
+
+private:
+    void on_save();
```
Use camel case like everything else.

jcelerier commented on tools/assetpipelineeditor/dracocompressor.cpp:
```diff
@@ -0,0 +1,166 @@
+                            attribute->vertexSize(),
+                            dracoDataType,
+                            false,
+                            attribute->byteStride(),
```
done

jcelerier commented:
@er-vin @lemirep  I added my last changes in a new commit to simplify review for you, when it's done I can split them in per-independent-class commits if you want (compressor, exporter, ui ?). However it's fairly hard to divide in commits for the various KUE-387 tasks, because if they aren't all done together then the export will produce meaningless gltf files.

jcelerier commented:
Finally fixed the export of the DodgeViper - joints indices weren't exported properly.

mkrus commented on src/core/gltf2importer/gltf2keys_p.h:
```diff
@@ -0,0 +1,88 @@
+
+    This file is part of Kuesa.
+
+    Copyright (C) 2018 Klarälvdalens Datakonsult AB, a KDAB Group company, info@kdab.com
```
2019

jcelerier replied:
done

mkrus commented on src/core/gltf2exporter/dracocompressor_p.cpp:
```diff
@@ -0,0 +1,314 @@
+
+QT_BEGIN_NAMESPACE
+Q_LOGGING_CATEGORY(dracocompressor, "DracoCompressor")
+namespace Kuesa::DracoCompression {
```
clang complains nested namespaces are a C++17 extension. We're targeting c++11

jcelerier replied:
done

mkrus commented on src/core/gltf2exporter/dracocompressor_p.h:
```diff
@@ -0,0 +1,69 @@
+namespace DracoCompression {
+
+struct CompressedMesh {
+    ~CompressedMesh();
```
why is the destructor needed? Adding it disables the generation of the default copy constructor which (with clang) leads to compile errors

mkrus replied:
weird, if I remove the destructor it builds for me. Stupid compilers.

Don't mind on the fix, but if you add the destructor then you need also add the copy constructor (at least, maybe the regular constructor too) for clang to be happy

jcelerier replied:
> weird, if I remove the destructor it builds for me. Stupid compilers.

I just checked and this looks like it's a matter of language version used - this does not work for me on gcc with C++11 either

jcelerier replied:
ah, the reason why it may have worked without destructor nor include, is because all the files that use dracocompressor_p.h also include the required draco header - but a new source file that would just do 

     #include "dracocompressor_p.h"
     void foo() { CompressedMesh mesh; } 

would fail to build

mkrus commented on src/core/gltf2exporter/gltf2exporter_p.cpp:
```diff
@@ -0,0 +1,716 @@
+        CompressedGLTFPrimitive compressed_primitive;
+
+        // Do the compression
+        const auto compressed = Kuesa::DracoCompression::compressMesh(*primitive.primitiveRenderer->geometry(), m_conf);
```
this fails to compile with clang if the destructor is added to the `CompressedMesh` class (which disables the generation of the default copy constructor)

mkrus commented on src/core/gltf2exporter/dracocompressor_p.cpp:
```diff
@@ -0,0 +1,314 @@
+
+template<>
+struct attribute_type_trait<char> {
+    static constexpr draco::DataType type = draco::DataType::DT_INT8;
```
clang warns that `type` is unused (as well as DT_UINT8, DT_INT16, DT_INT32, DT_UINT32, DT_FLOAT64).

mkrus commented on src/core/gltf2exporter/dracocompressor_p.cpp:
```diff
@@ -0,0 +1,314 @@
+template<typename T>
+void addFacesToMesh(const QVector<T> &indices, draco::Mesh &dracoMesh)
+{
+    using index_type = draco::IndexType<unsigned int, draco::PointIndex_tag_type_>;
```
why this? Looks like `Face` is just an array of `PointIndex`?

mkrus commented on tools/assetpipelineeditor/exportdialog.ui:
```diff
@@ -0,0 +1,314 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
could you remove the enabled and make the `QGroupBox` checkable?

jcelerier replied:
done

mkrus commented on tools/assetpipelineeditor/mainwindow.cpp:
```diff
@@ -51,7 +52,14 @@
+#include <GLTF2Importer>
+
 #include <Kuesa/SceneEntity>
+#include <private/gltf2exporter_p.h>
```
`Kuesa/private/...` I think is needed

jcelerier replied:
done

mkrus commented on tools/assetpipelineeditor/mainwindow.ui:
```diff
@@ -131,7 +134,7 @@
   <widget class="QDockWidget" name="animationDockWidget">
    <property name="windowTitle">
-    <string>Animations</string>
+    <string>A&amp;nimations</string>
```
is A used? would make more sense...

jcelerier replied:
it's used for "About" but maybe that one is less relevant

jcelerier commented:
hmm I don't know why Sean's commit got in the middle, gonna rebase this

jcelerier commented:
@er-vin do you want me to squash and recreate atomic commits for the various features from this ?

er-vin commented on src/core/gltf2exporter/dracocompressor_p.h:
```diff
@@ -0,0 +1,69 @@
+    along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#ifndef KUESA_GLTF2EXPORTER_DRACOCOMPRESSOR_H
```
Nitpick: should end with _P_H

jcelerier replied:
done

er-vin commented on src/core/gltf2exporter/gltf2exporter_p.h:
```diff
@@ -0,0 +1,125 @@
+/*
+    gltf2exporter.h
```
gltf2exporter_p.h

jcelerier replied:
done

er-vin commented on src/core/gltf2exporter/gltf2exporter_p.h:
```diff
@@ -0,0 +1,125 @@
+    along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#ifndef KUESA_GLTF2EXPORTER_GLTF2EXPORTER_H
```
Should use _P_H

jcelerier replied:
done

er-vin commented on tests/auto/gltfexporter/tst_gltfexporter.cpp:
```diff
@@ -0,0 +1,198 @@
+                    QDir(ASSETS),
+                    tmp);
+        }
+    }
```
Can't we at least check something to verify the compression occured? It's lacking some QCOMPARE or QVERIFY in there. :-)

jcelerier replied:
done

er-vin commented on tools/assetpipelineeditor/exportdialog.cpp:
```diff
@@ -0,0 +1,104 @@
+                QMessageBox::Cancel);
+        if (res != QMessageBox::Ok) {
+            ui->errorLog->setPlainText(tr("Nothing saved."));
+            return;
```
Unsure since I couldn't try it yet (I lack draco), but isn't it going to close the dialog still when getting there? If yes this should be fixed.

jcelerier replied:
hmm, on my machine (linux) the export dialog isn't closed after saving, but maybe this is something platform-specific - I will try on mac & windows

er-vin replied:
OK, just me being confused on save accepting the dialog or not. I was surely confused, feel free to ignore this.

er-vin commented on tools/assetpipelineeditor/exportdialog.cpp:
```diff
@@ -0,0 +1,104 @@
+{
+    ui->errorLog->setPlainText("");
+
+    m_targetFile = QFileDialog::getSaveFileName(this, tr("Select a folder to save to"), m_targetFile, tr("glTF 2.0 file (*.gltf)"));
```
We're not handling the case where the user clicked cancel and we got an empty string back. We should likely bail out early then. Would be even better if that was selected from within the original dialog instead of a file dialog on top of a dialog when clicking save.

jcelerier replied:
is the best way to go towards this is to make the ExportDialog inherit from QFileDialog and add the compression controls at the bottom ? (and is it ok if it's not the native file dialogs anymore in that case ?)

jcelerier replied:
also it used to be like this before, should I maybe revert to that ? https://user-images.githubusercontent.com/2772730/51386615-57104480-1b23-11e9-8095-e198d147c6b7.png

er-vin replied:
Yes, I think I preferred that older approach indeed. Sorry for the rework incurred.

mkrus replied:
old approach had the issue that it didn't warm the user about overwriting the original file

er-vin replied:
Clearly it should warn of course, this is unrelated to how the GUI though.

jcelerier replied:
![mesh](https://user-images.githubusercontent.com/2772730/52701296-300a2e80-2f7a-11e9-9d15-3df0c7eec5ea.png) here's how it looks now - the icon on the lineedit opens the save dialog. Is it ok ? (it seemed clearer to me than the previous solution)

er-vin commented on tools/assetpipelineeditor/mainwindow.cpp:
```diff
@@ -337,6 +350,21 @@ void MainWindow::openFile()
     setFilePath(filePath);
 }
 
+void MainWindow::compressFile()
```
Probably shouldn't be name compress since it'll be more generally about exporting.

jcelerier replied:
done

jcelerier replied:
done

er-vin commented on tools/assetpipelineeditor/mainwindow.cpp:
```diff
@@ -164,6 +173,10 @@ void MainWindow::updateScene(Kuesa::SceneEntity *entity)
     m_ui->collectionBrowser->expandAll();
     m_ui->actionReload->setEnabled(true);
+#if defined(KUESA_DRACO_COMPRESSION)
+    m_ui->actionExport->setEnabled(true);
```
I'd go as far as completely hiding it from the GUI for now as long as it's useful only in the draco case. Will change once textures can be marked for post-processing too.

jcelerier replied:
done

jcelerier commented:
just pushed with the changes requested - here's the diff attached if it can make it simpler to review (I also added a safeguard for the created name of the compressed binary file - if there is already one with this name in the source folder it will try to use another - compressedBuffer-1.bin, etc) : 
[out.txt](https://github.com/KDAB/kuesa/files/2859690/out.txt)
